streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -133,34 +141,37 @@ public void process(final K key, final V1 value) {
                 return;
             }
 
-            // maxObservedStreamTime is updated and shared between left and right sides, so we can
-            // process a non-join record immediately if it is late
-            final long maxStreamTime = maxObservedStreamTime.updateAndGet(time -> Math.max(time, context().timestamp()));
-
             boolean needOuterJoin = outer;
+            boolean joinFound = false;
 
             final long inputRecordTimestamp = context().timestamp();
             final long timeFrom = Math.max(0L, inputRecordTimestamp - joinBeforeMs);
             final long timeTo = Math.max(0L, inputRecordTimestamp + joinAfterMs);
 
+            // maxObservedStreamTime is updated and shared between left and right join sides
+            maxObservedStreamTime.updateAndGet(t -> Math.max(t, inputRecordTimestamp));
+
             try (final WindowStoreIterator<V2> iter = otherWindow.fetch(key, timeFrom, timeTo)) {
                 while (iter.hasNext()) {
                     needOuterJoin = false;
+                    joinFound = true;
                     final KeyValue<Long, V2> otherRecord = iter.next();
                     final long otherRecordTimestamp = otherRecord.key;
+
+                    // Emit expired records before the joined record to keep time ordering
```

Comment:
I think we can refactor the logic here as the following:

0) suppose the received record timestamp is T1, the current stream time is T2 >= T1; and we found one or more matching record from the other side, with timestamp T1' <= T2' <= T3' etc. The joined record would have the timestamp of T1` = max(T1, T1'), T2` = max(T1, T2'), where T1` <= T2` <= ...

1) After we get all the joined records, we do not call `context.forward()` yet, but just cache them locally.

2) We then range query the expired records store, and generate the joined records (and also delete the records), again we do not call `context.forward()` yet, but just cache them locally. 

3) We merge sort on these two sorted-by-timestamp list, and then call `context.forward()` on the sorted join result records to emit.

In this we do not need the following complex logic.

Reply:
cc @mjsax as well, LMK WDYT.

Reply:
I guess that's possible, but _if_ the join result is large, we could run into memory issue buffering all join results?

Also, sorting could be expensive and we can actually avoid it, and still guarantee that results are emitted in timestamp order:
- we know that left/outer join result would have the smallest timestamps and thus we can emit those first (given that we use timestamped-sorted store anyway, we just scan the store from old to new and emit
- for the inner join result, we get the output sorted by timestamp, too, because for the join key, data is sorted in timestamp order in the store, too

Reply:
The trickiness as demonstrated in the current PR though, is that if we first do the expiration we may get records that are matched to the current processing record, which need to be skipped from deleting/emitting before the join.

I think it is still possible to simply the current logic without naive buffering. Because:

1) The current processing record's timestamp T is no larger than the updated max stream time T';
2) The current processing record's matching record's smallest timestamp would be (T - window-size);
3) The expired records' largest timestamp would be (T' - window-size - grace-period), where grace-period >= 0.

In other words, if the current processing's record timestamp T is smaller than T' (i.e. it's a late record and hence did not advance the stream time), then for all records that are within [T - window-size, T' - window-size - grace-period] assuming T - window-size < T' - window-size - grace-period, would have already been expired end emitted, and hence won't be found and matched; if the current processing's record timestamp T == T' (i.e. it is not a late record), then T - window-size is always >= T' - window-size - grace-period, which means that all joined record's timestamps should be later than the expired timestamps. 

That means, if we do the expiration first based on (T' - window-size - grace-period), the newly expired records' timestamps should all be smaller than any joined record's timestamps for that processing record generated later. And hence it is safe to just blindly expire them all without the `except` logic.

Reply:
I agree (even if your train of though seems to be complex...) that we can expired left/outer join result outside of the loop blindly.

To rephrase: if we process record with timestamp T there are two cases:

In-order record:
We can expire everything with timestamp smaller then `T - windowSize - gracePeriod`.

Out-of-order record:
There is nothing to expired to begin with, because we could only expire `T - windowSize - gracePeriod` but this value is smaller than `streamTime - windowSize - gracePeriod` and thus we expired everything up to the larger value perviously already.

Ie, we don't need method `emitExpiredNonJoinedOuterRecordsExcept` at all, and can move `emitExpiredNonJoinedOuterRecords();` before the while loop.